### PR TITLE
Allow configuration of Kafka topics with periods

### PR DESCRIPTION
**Pull request by rschu1ze**
The Kafka table engine allows global configuration and per-Kafka-topic configuration. The latter uses syntax `<kafka_TOPIC>`, e.g. for topic "football":

```xml
  <kafka>
    <!-- global configuration ... -->
  </kafka>

  <kafka_football>
      <retry_backoff_ms>250</retry_backoff_ms>
      <fetch_min_bytes>100000</fetch_min_bytes>
  </kafka_football>
```

Some users had to find out the hard way that such configuration doesn't take effect if the topic name contains a period, e.g. "sports.football". The reason is that ClickHouse configuration framework already uses periods as level separators to descend the configuration hierarchy. (Besides that, per-topic configuration at the same level as global configuration could be considered ugly.) Note that Kafka topics may contain characters "a-zA-Z0-9._-" (*) and a tree-like topic organization using periods is quite common in practice.

This PR deprecates the existing per-topic configuration syntax (but continues to support it for backward compat) and introduces a new per-topic configuration syntax below the global Kafka configuration of the form:

```xml
<kafka>
   <kafka_topic>
       <name>football</name>
       <retry_backoff_ms>250</retry_backoff_ms>
       <fetch_min_bytes>100000</fetch_min_bytes>
   </kafka_topic>
   <kafka_topic>
       <name>baseball</name>
       <retry_backoff_ms>400</retry_backoff_ms>
       <fetch_min_bytes>50000</fetch_min_bytes>
   </kafka_topic>
</kafka>
```

The period restriction no longer applies, so `<name>sports.football</name>` will work. Also, everything Kafka-related is below `<kafka>` which is 13% aesthetically more pleasing than before. And, as a bonus, per-topic configuration is structurally equivalent to the way [S3 endpoints are configured](https://clickhouse.com/docs/en/integrations/s3/s3-table-engine/#managing-credentials).

Considered but rejected alternatives:
- Extending Poco ConfigurationView with custom separators (e.g."/" instead of "."). Won't work easily because ConfigurationView only builds a path but defers descending the configuration tree to the normal configuration classes. We would be back at square one.
- Reloading the configuration file in StorageKafka (instead of reading from the already loaded file) but with a custom separator. This mode is supported by XML configuration, not sure about YAML. Too ugly and too error-prone since the internal configuration is composed from multiple configuration files (think user.xml).

(*) https://stackoverflow.com/a/37067544

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
It is now possible using new configuration syntax to configure Kafka topics with periods in their name

### CODE REVIEW

**Comment by antonio2368** [comment]
Why not use `<topic>` key for topic name? If it's defined it will be preferred over `<kafka_TOPIC>` one.

**Comment by antonio2368** [comment]
Check how we can define credentials https://clickhouse.com/docs/en/integrations/s3/s3-table-engine/#managing-credentials

We define key `<endpoint-name>` but don't use it, all the information is contained inside the section itself.

In the same way `<kafka_some_name>` would be used to denote that the following is kafka configuration + it seems easier to achieve backward compatibility.
Additionally, we can do the same as for S3 and support:

```
<kafka>
   <some-unique-name>
       <topic></topic>
       ...
   </some-unique-name>
</kafka>
```

Also, how do attributes work with yaml?

**Comment by rschu1ze** [comment]
Implemented now things the way you proposed.

**Comment by rschu1ze** [comment]
ClickHouse Integration Tests (asan) [5/6]:
- test_tcp_handler_interserver_listen_host: "Pool overlaps with other one on this address space" - infra issue?

ClickHouse Stress Test (ubsan):
ClickHouse Stress Test (debug):
- The many crashes (e.g. with query `SELECT sum(cityHash64(*)`) are unrelated, have seen them elsewhere too.

File: src/Storages/Kafka/StorageKafka.cpp [hunk 1]
```diff
@@ -175,22 +175,69 @@ namespace
     const auto CLEANUP_TIMEOUT_MS = 3000;
     const auto MAX_THREAD_WORK_DURATION_MS = 60000;  // once per minute leave do reschedule (we can't lock threads in pool forever)
 
-    /// Configuration prefix
-    const String CONFIG_PREFIX = "kafka";
+    const String CONFIG_KAFKA_TAG = "kafka";
+    const String CONFIG_KAFKA_TOPIC_TAG = "kafka_topic";
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Storages/Kafka/StorageKafka.cpp:179]: I had it that way (`topic`) first. Problem is that there exist [kafka settings](https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md) starting with `topic`. The XML tags (and their yaml equivalents) in ClickHouse's configuration file are translated into these settings. So meta-tag `topic` would potentially lead to clashes. We need something unique.

There are no Kafka setting `kafka.topic` (corresponding to `<kafka_topic>`) and I don't think they will add such settings in future (as "kafka" prefix is redundant from their POV). In sum: `<kafka_topic>` isn't super-beautiful but it does the job as a meta-tag.

File: src/Storages/Kafka/StorageKafka.cpp [hunk 2]
```diff
@@ -175,22 +175,69 @@ namespace
...
+            const String setting_name_in_kafka_config = (tag == "log_level") ? tag : boost::replace_all_copy(tag, "_", ".");
+            kafka_config.set(setting_name_in_kafka_config, setting_value);
+        }
+    }
+
+    /// Read server configuration into cppkafa configuration, used by new per-topic configuration
+    void loadTopicConfig(cppkafka::Configuration & kafka_config, const Poco::Util::AbstractConfiguration & config, const String & config_prefix, const String & topic)
+    {
+        /// Read all tags one level below <kafka>
+        Poco::Util::AbstractConfiguration::Keys tags;
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Storages/Kafka/StorageKafka.cpp:208]: Good idea but overkill here as the configuration happens only when a new Kafka table engine is created resp. before the first write as far as I see.

File: docs/en/engines/table-engines/integrations/kafka.md [hunk 3]
```diff
@@ -162,22 +162,59 @@ If you want to change the target table by using `ALTER`, we recommend disabling
...
+  <!-- Configuration specific to topics "logs" and "stats" -->
+  <!-- Does NOT support periods in topic names, e.g. "logs.security"> -->
+
   <kafka_logs>
     <retry_backoff_ms>250</retry_backoff_ms>
     <fetch_min_bytes>100000</fetch_min_bytes>
   </kafka_logs>
+
+  <kafka_stats>
+    <retry_backoff_ms>400</retry_backoff_ms>
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/engines/table-engines/integrations/kafka.md:210]: That's the legacy syntax (l. 192) ... we better don't touch it.

File: src/Storages/Kafka/StorageKafka.cpp [hunk 4]
```diff
@@ -175,22 +175,69 @@ namespace
...
+        config.keys(config_prefix, tags);
+
+        for (const auto & tag : tags)
+        {
+            /// Consider tag <kafka_topic>. Multiple occurrences given as "kafka_topic", "kafka_topic[1]", etc.
+            if (tag.starts_with(CONFIG_KAFKA_TOPIC_TAG))
```
Comment (antonio2368) [inline src/Storages/Kafka/StorageKafka.cpp:214]: early continue if it doesn't start with `CONFIG_KAFKA_TOPIC_TAG` to eliminate indentation

File: src/Storages/Kafka/StorageKafka.cpp [hunk 5]
```diff
@@ -543,21 +594,47 @@ void StorageKafka::updateConfiguration(cppkafka::Configuration & conf)
...
     {
         const auto topic_config_key = config_prefix + "_" + topic;
         if (config.has(topic_config_key))
-            loadFromConfig(conf, config, topic_config_key);
+            loadFromConfig(kafka_config, config, topic_config_key);
     }
 
+    // Update consumer topic-specific configuration (new syntax). Example with topics "football" and "baseball":
+    //     <kafka>
+    //         <topic>
```
Comment (antonio2368) [inline src/Storages/Kafka/StorageKafka.cpp:619]:
```suggestion
    //         <kafka_topic>
```

**Review by antonio2368 — APPROVED** [review]

File: src/Storages/Kafka/StorageKafka.cpp (hunk 5 above)
Comment (rschu1ze) [reply src/Storages/Kafka/StorageKafka.cpp]: Thanks, I knew I forgot something somewhere ... :-)

File: src/Storages/Kafka/StorageKafka.cpp (hunk 4 above)
Comment (rschu1ze) [reply src/Storages/Kafka/StorageKafka.cpp:214]: Done

**Pull request merged**